ui: Liveness info on Nodes List page

mrtracy opened the pull request:
Two main additions:
+ The Admin UI now queries Liveness data from the backend, and can now use it to better
inform status visuals across the cluster. This data is added to the Redux state;
additionally, numerous selectors have been created in the new `redux/nodes.ts`
which provide useful computed information to various components.
+ This information is now being used by the Nodes List to divide all nodes into
"Live" and "Dead".

The nodes page has been changed to display two tables - one for live nodes, and
another for dead nodes. This fixes #5415, as dead nodes now have a different set
of columns than live nodes and thus we are no longer displaying out-of-date
values for these nodes. This also fixes #12422, as dead nodes are no longer
considered in summation totals. Also makes progress towards #13895.

While overhauling the nodes page, have also added a tooltip to the colored
health indicators on the nodes list to help explain the "healthy", "suspect" and
"dead" states. Fixes #7547.

Division of nodes list:
<img width="1305" alt="screen shot 2017-03-19 at 7 11 34 pm" src="https://cloud.githubusercontent.com/assets/6969858/24085771/246cfbd2-0cd8-11e7-9735-0368c9faeb8e.png">


Tooltip:
<img width="865" alt="screen shot 2017-03-19 at 7 05 06 pm" src="https://cloud.githubusercontent.com/assets/6969858/24085774/28c63cac-0cd8-11e7-9507-4a1c6709acb8.png">

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14258)
<!-- Reviewable:end -->

Commit:
ui: Liveness info on Nodes List page

Two main additions:
+ The Admin UI now queries Liveness data from the backend, and can now use it to better
inform status visuals across the cluster. This data is added to the Redux state;
additionally, numerous selectors have been created in the new `redux/nodes.ts`
which provide useful computed information to various components.
+ This information is now being used by the Nodes List to divide all nodes into
"Live" and "Dead".

The nodes page has been changed to display two tables - one for live nodes, and
another for dead nodes. This fixes #5415, as dead nodes now have a different set
of columns than live nodes and thus we are no longer displaying out-of-date
values for these nodes. This also fixes #12422, as dead nodes are no longer
considered in summation totals. Also makes progress towards #13895.

While overhauling the nodes page, have also added a tooltip to the colored
health indicators on the nodes list to help explain the "healthy", "suspect" and
"dead" states. Fixes #7547.

tamird commented:
Reviewed 8 of 9 files at r1.
Review status: 8 of 9 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/ui/app/redux/cachedDataReducer.ts, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kfd6mS-ZeuOyrmq7CbC:-Kfd6mS-ZeuOyrmq7CbD:bz62vge) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb320b53a9f8e85958f4cf24cd2ec57677fd31ec/pkg/ui/app/redux/cachedDataReducer.ts#L41)):*
> ```XML
>   // Track all the currently seen namespaces, to ensure there isn't a conflict
>   private static namespaces: { [actionNamespace: string]: boolean } = {};
> ```

why remove?

---

*[pkg/ui/app/redux/nodes.ts, line 12 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kfd7QCuetmxbqRsHCUv:-Kfd7QCuetmxbqRsHCUw:b-n137be) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c7b9a427db1d1913bcc3ece6f400b133b5fa7da/pkg/ui/app/redux/nodes.ts#L12)):*
> ```XML
> // considered to be alive; if more than this duration has passed since the
> // node's liveness expired, it is considered dead.
> export const deadTimeout = moment.duration(5, "m");
> ```

this needs to be kept in sync with some backend constant, yeah? ditto below i think.

---

*[pkg/ui/app/redux/nodes.ts, line 33 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kfd7zVvXaQMxHjV8vl3:-Kfd7zVvXaQMxHjV8vl4:bkjla4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c7b9a427db1d1913bcc3ece6f400b133b5fa7da/pkg/ui/app/redux/nodes.ts#L33)):*
> ```XML
> 
> // Functions to select data directly from the redux state.
> let nodeStatuses = (state: AdminUIState) => state.cachedData.nodes.data;
> ```

throughout: consts?

---

*[pkg/ui/app/redux/nodes.ts, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kfd7jW90tOh6oFiGuW5:-Kfd7jW90tOh6oFiGuW6:b-u6frkw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c7b9a427db1d1913bcc3ece6f400b133b5fa7da/pkg/ui/app/redux/nodes.ts#L41)):*
> ```XML
> let livenessByNodeID = createSelector(
>   livenesses,
>   (live) => {
> ```

s/live/liveness/? livenessRecord?

---

*[pkg/ui/app/redux/nodes.ts, line 56 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kfd7whHCsnOUW4tgyXX:-Kfd7whHCsnOUW4tgyXY:bomjwqq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c7b9a427db1d1913bcc3ece6f400b133b5fa7da/pkg/ui/app/redux/nodes.ts#L56)):*
> ```XML
>   (liveByNodeID, liveCheckedAt) => {
>     if (liveCheckedAt) {
>       let deadCutoff = liveCheckedAt.clone().subtract(deadTimeout);
> ```

lots of easy `const`s here.

---

*[pkg/ui/app/redux/nodes.ts, line 62 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kfd7rnEjvrweLWhIFxU:-Kfd7rnEjvrweLWhIFxV:b-hddbko) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c7b9a427db1d1913bcc3ece6f400b133b5fa7da/pkg/ui/app/redux/nodes.ts#L62)):*
> ```XML
>         let status = LivenessStatus.HEALTHY;
>         if (expiration.isBefore(deadCutoff)) {
>           status = LivenessStatus.DEAD;
> ```

why all this mutability? you can just return in each arm.

---

*[pkg/ui/app/redux/nodes.ts, line 97 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kfd8C17OnDwiALOYbcZ:-Kfd8C17OnDwiALOYbc_:byb5ghc) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c7b9a427db1d1913bcc3ece6f400b133b5fa7da/pkg/ui/app/redux/nodes.ts#L97)):*
> ```XML
> // nodeSums returns an object with certain cluster-wide totals which are used
> // in different places in the UI.
> let nodeSums = createSelector(
> ```

this looks like it could use a test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 5 of 9 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/ui/app/redux/cachedDataReducer.ts, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kfd6mS-ZeuOyrmq7CbC:-Kfh-c9todWs_HNdBXyE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb320b53a9f8e85958f4cf24cd2ec57677fd31ec/pkg/ui/app/redux/cachedDataReducer.ts#L41)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why remove?
</blockquote></details>

Done.

---

*[pkg/ui/app/redux/nodes.ts, line 12 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kfd7QCuetmxbqRsHCUv:-Kfh6UHUBwxbZfK1s0Df:bfm4xu7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c7b9a427db1d1913bcc3ece6f400b133b5fa7da/pkg/ui/app/redux/nodes.ts#L12)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this needs to be kept in sync with some backend constant, yeah? ditto below i think.
</blockquote></details>

Updated this comment; it should be kept in sync with a value for now, but when #14230 is complete the configured value should be pulled from the server.

---

*[pkg/ui/app/redux/nodes.ts, line 33 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kfd7zVvXaQMxHjV8vl3:-Kfh8KOVl2iYDFGfBmFm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c7b9a427db1d1913bcc3ece6f400b133b5fa7da/pkg/ui/app/redux/nodes.ts#L33)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

throughout: consts?
</blockquote></details>

Done.

---

*[pkg/ui/app/redux/nodes.ts, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kfd7jW90tOh6oFiGuW5:-Kfh8gHh7Xn30Hs2rJKm:b-r4nknb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c7b9a427db1d1913bcc3ece6f400b133b5fa7da/pkg/ui/app/redux/nodes.ts#L41)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/live/liveness/? livenessRecord?
</blockquote></details>

livenessRecord

---

*[pkg/ui/app/redux/nodes.ts, line 56 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kfd7whHCsnOUW4tgyXX:-Kfh8ioFhCsjI_sQQ0hs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c7b9a427db1d1913bcc3ece6f400b133b5fa7da/pkg/ui/app/redux/nodes.ts#L56)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

lots of easy `const`s here.
</blockquote></details>

Done.

---

*[pkg/ui/app/redux/nodes.ts, line 62 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kfd7rnEjvrweLWhIFxU:-Kfh8jfHC29dZNBEoLn4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c7b9a427db1d1913bcc3ece6f400b133b5fa7da/pkg/ui/app/redux/nodes.ts#L62)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why all this mutability? you can just return in each arm.
</blockquote></details>

Done.

---

*[pkg/ui/app/redux/nodes.ts, line 97 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kfd8C17OnDwiALOYbcZ:-Kfhtx4U-dnSjK6ac3JL:b-96evlg) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c7b9a427db1d1913bcc3ece6f400b133b5fa7da/pkg/ui/app/redux/nodes.ts#L97)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this looks like it could use a test.
</blockquote></details>

It could, as could a lot of stuff around our selectors; unfortunately it is deceptively difficult to set up the fake redux state we need to test this selector correctly.  I don't think a test for this will be worth the effort to create it within the scope of this pull request.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258)*
<!-- Sent from Reviewable.io -->

tamird commented:
I'll let @maxlang sign off on this.

---

Reviewed 3 of 3 files at r2.
Review status: 8 of 9 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/ui/app/containers/nodesOverview.tsx, line 57 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-KfiF8q8QYb3jzxvenT_:-KfiF8q8QYb3jzxvenTa:bykc78m) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/containers/nodesOverview.tsx#L57)):*
> ```XML
> 
>   render() {
>     let { statuses, nodesSummary, sortSetting } = this.props;
> ```

looks like a lot of easy `const`s in the changes made to this file as well.

---

*[pkg/ui/app/redux/nodes.ts, line 212 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-KfiEo-Bed9nMdBAAIXO:-KfiEo-Bed9nMdBAAIXP:b-ds5jav) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/redux/nodes.ts#L212)):*
> ```XML
>     return null;
> }
> let nodesSummaryType = getReturnType(nodesSummary);
> ```

can this be const, too? can it be inlined below?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-:-KfiFCo_Nbkfkr7UX8R9:b-cx7qkp)*
<!-- Sent from Reviewable.io -->

benesch commented:
This looks awesome, @mrtracy. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, but would love to get your thoughts on a few comments below.

---

Reviewed 6 of 9 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/ui/app/containers/nodesOverview.tsx, line 82 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-KfxH7uiW7cuz7dcmqql:-KfxH7uiW7cuz7dcmqqm:bqo45bo) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/containers/nodesOverview.tsx#L82)):*
> ```XML
>                   let tooltip = (status === LivenessStatus.HEALTHY) ?
>                     "This node is currently healthy." :
>                     `This node has not reported as being live for over ${suspectTimeout.humanize()}`;
> ```

Inconsistent punctuation here.

---

*[pkg/ui/app/redux/cachedDataReducer.spec.ts, line 226 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kg0971lHnZH5cvcvaTd:-Kg0971lHnZH5cvcvaTe:byuqcyd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/redux/cachedDataReducer.spec.ts#L226)):*
> ```XML
>         expected[id].setAt = testMoment;
>         assert.deepEqual(state, expected);
>       });
> ```

:thumbsup: to this whole file

---

*[pkg/ui/app/redux/cachedDataReducer.ts, line 10 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kg07O0m-iGIomZ1PEH4:-Kg07O0m-iGIomZ1PEH5:bd6thzo) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6e665dff7bcdd86907ee35c91f8101d4c1904b7/pkg/ui/app/redux/cachedDataReducer.ts#L10)):*
> ```XML
> import { Dispatch } from "redux";
> import { assert } from "chai";
> import moment = require("moment");
> ```

I'm kind of surprised this worked.

---

*[pkg/ui/app/redux/nodes.ts, line 63 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kg-jehmFhBshWFjLD_c:-Kg-jehmFhBshWFjLD_d:b-z9c1ih) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/redux/nodes.ts#L63)):*
> ```XML
> const livenessByNodeID = createSelector(
>   livenesses,
>   (livenessRecord) => {
> ```

Oh my goodness, does the linter make you rename these parameters to not match their real names? That is excruciating.

---

*[pkg/ui/app/redux/nodes.ts, line 81 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kg08Gtv21l6obMSG0vV:-Kg08Gtv21l6obMSG0vW:b5abam4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/redux/nodes.ts#L81)):*
> ```XML
>     if (liveCheckedAt) {
>       const deadCutoff = liveCheckedAt.clone().subtract(deadTimeout);
>       const suspectCutoff = liveCheckedAt.clone().subtract(suspectTimeout);
> ```

Not for this PR, but wanted to put this on your radar: we should consider switching to WhoopInc/frozen-moment—a version of MomentJS that's immutable by default. Moment is probably going immutable in mainline eventually (moment/moment#3548).

---

*[pkg/ui/app/redux/nodes.ts, line 102 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-KfxJCTQ3GWGcbpDPdHd:-KfxJCTQ3GWGcbpDPdHe:bpt6bmd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/redux/nodes.ts#L102)):*
> ```XML
>   nodeStatuses,
>   (nss) => {
>     return _.map(nss, (ns) => {
> ```

Any reason you're using `_.map` instead of `nss.map` here? Asking mostly for my own sake.

---

*[pkg/ui/app/redux/nodes.ts, line 211 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-KfxJxxrh4g1xDRMC03d:-KfxJxxrh4g1xDRMC03e:b-bzmd7d) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/redux/nodes.ts#L211)):*
> ```XML
> function getReturnType<R> (_: (...args: any[]) => R): R {
>     return null;
> }
> ```

Handy! Seems like it belongs in a util file though?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-:-Kg0A6QWwXTZ_to1lxv3:br9re8p)*
<!-- Sent from Reviewable.io -->

maxlang commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/ui/app/containers/nodesOverview.tsx, line 175 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kg1RHKzNfbKxvItBeLH:-Kg1RHKzNfbKxvItBeLI:b-z14uer) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/containers/nodesOverview.tsx#L175)):*
> ```XML
>       <section className="header header--subsection">
>         Dead Nodes
>       </section>
> ```

I know this is what we're currently doing but it feels like this should be an `h1` or something not a `section`. Feels like we're misusing html5.

---

*[pkg/ui/app/containers/nodesOverview.tsx, line 190 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kg1RkhfXxa3eSPbxLwg:-Kg1RkhgJvkaFk52Uyiw:b-f63ahs) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/containers/nodesOverview.tsx#L190)):*
> ```XML
>                   return <div>
>                     <Link to={"/nodes/" + ns.desc.node_id}>{ns.desc.address.address_field}</Link>
>                     <div className={"icon-circle-filled node-status-icon node-status-icon--dead"}
> ```

Why is this surrounded by curly braces? Can't it just be used directly as a string?

---

*[pkg/ui/app/containers/nodesOverview.tsx, line 208 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kg1S43tXp63Br24h8WM:-Kg1S43tXp63Br24h8WN:bl09of6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/containers/nodesOverview.tsx#L208)):*
> ```XML
>                   let liveness = nodesSummary.livenessByNodeID[ns.getDesc().getNodeId()];
>                   let deadTime = liveness.expiration.getWallTime().toNumber();
>                   let deadMoment = moment(NanoToMilli(deadTime));
> ```

The `LongToMoment` util function in `convert.ts` would let you save a couple lines here.

---

*[pkg/ui/app/redux/nodes.ts, line 63 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kg-jehmFhBshWFjLD_c:-Kg1TQTzbYdrMEq0UPkg:b-rt06b2) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/redux/nodes.ts#L63)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Oh my goodness, does the linter make you rename these parameters to not match their real names? That is excruciating.
</blockquote></details>

I think a good pattern going forward would be to call the various accessor/selector functions something like `livenessesSelector` and then the argument `livenesses`.

---

*[pkg/ui/app/redux/nodes.ts, line 81 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kg08Gtv21l6obMSG0vV:-Kg1U-8SMOJsRI_lVEGq:b-jpc2s2) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/redux/nodes.ts#L81)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Not for this PR, but wanted to put this on your radar: we should consider switching to WhoopInc/frozen-moment—a version of MomentJS that's immutable by default. Moment is probably going immutable in mainline eventually (moment/moment#3548).
</blockquote></details>

Man this probably would've saved me a lot of pain. Especially in my angular days.

---

*[pkg/ui/app/redux/nodes.ts, line 102 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-KfxJCTQ3GWGcbpDPdHd:-Kg1UVVPfaL7HjPAz-So:b-aqy041) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/redux/nodes.ts#L102)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Any reason you're using `_.map` instead of `nss.map` here? Asking mostly for my own sake.
</blockquote></details>

Can't speak for @mrtracy but I can say my trust in js is so low I personally just opt to use `lodash` because I expect it'll be backwards compatible and fast. If the native implementation is fastest it'll just use that under the hood. Also it just makes operations like this read more consistently because `lodash` `each` and `map` operate similarly over arrays and objects.

I hadn't looked this up before but just to further support this argument: `lodash` is faster than native `map` but slower than a `for` loop....
https://jsperf.com/native-map-vs-lodash-map/21

---

*[pkg/ui/app/redux/nodes.ts, line 103 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kg1WOqhpPQz60XTvz6j:-Kg1WOqhpPQz60XTvz6k:bjf4dai) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/redux/nodes.ts#L103)):*
> ```XML
>   (nss) => {
>     return _.map(nss, (ns) => {
>       return ns.desc.node_id.toString();
> ```

This function could be a one liner
```
(ns) => ns.desc.node_id.toString()
```

---

*[pkg/ui/app/redux/nodes.ts, line 199 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kg1XRqZrxOyQjWCOBUF:-Kg1XRqZrxOyQjWCOBUG:b-qjrjn7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/redux/nodes.ts#L199)):*
> ```XML
>     };
>   },
> );
> ```

There's probably a more succinct way of doing this using weird new JS stuff but I can't think of anything off the top of my head.

---

*[pkg/ui/app/redux/nodes.ts, line 211 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-KfxJxxrh4g1xDRMC03d:-Kg1Z5FAns82XZ0hRXm4:blqggkp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/redux/nodes.ts#L211)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Handy! Seems like it belongs in a util file though?
</blockquote></details>

Cool idea

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-:-Kg1ZPC3i0l-EhxISrW6:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 1 of 10 files reviewed at latest revision, 14 unresolved discussions.

---

*[pkg/ui/app/containers/nodesOverview.tsx, line 57 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-KfiF8q8QYb3jzxvenT_:-KgG1JNdFk81fO7K5OUG:bdeb7jb) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/containers/nodesOverview.tsx#L57)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

looks like a lot of easy `const`s in the changes made to this file as well.
</blockquote></details>

I was able to `s/let/const/` through this file, i'll try to do this on other files when I touch them for code reviews.

---

*[pkg/ui/app/containers/nodesOverview.tsx, line 82 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-KfxH7uiW7cuz7dcmqql:-KgGCnfk6cSjs_Xj-TTG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/containers/nodesOverview.tsx#L82)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Inconsistent punctuation here.
</blockquote></details>

Done.

---

*[pkg/ui/app/containers/nodesOverview.tsx, line 175 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kg1RHKzNfbKxvItBeLH:-KgG1WuZ8yC4pLSWEFBL:b-7debg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/containers/nodesOverview.tsx#L175)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary><blockquote>

I know this is what we're currently doing but it feels like this should be an `h1` or something not a `section`. Feels like we're misusing html5.
</blockquote></details>

I'm not sure what the best thing to is here. I'm converting it to a div for now, i've been planning to do a once-over before 1.0 to get this kind of thing right.

---

*[pkg/ui/app/containers/nodesOverview.tsx, line 190 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kg1RkhfXxa3eSPbxLwg:-KgG2C5q1EWm2U6j7rVV:b-tr8en8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/containers/nodesOverview.tsx#L190)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary><blockquote>

Why is this surrounded by curly braces? Can't it just be used directly as a string?
</blockquote></details>

It's companion in LiveNodes needs curly braces, a simple copy-paste artifact. Fixed.

---

*[pkg/ui/app/containers/nodesOverview.tsx, line 208 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kg1S43tXp63Br24h8WM:-KgG2ZXJ9aLH0VAY5LL2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/containers/nodesOverview.tsx#L208)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary><blockquote>

The `LongToMoment` util function in `convert.ts` would let you save a couple lines here.
</blockquote></details>

Done.

---

*[pkg/ui/app/redux/cachedDataReducer.spec.ts, line 226 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kg0971lHnZH5cvcvaTd:-KgGCj5v8czc6rLmV0VO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/redux/cachedDataReducer.spec.ts#L226)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

:thumbsup: to this whole file
</blockquote></details>

Done.

---

*[pkg/ui/app/redux/cachedDataReducer.ts, line 10 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kg07O0m-iGIomZ1PEH4:-KgG2bMn0cUNAXEi-SyQ:b1170bs) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6e665dff7bcdd86907ee35c91f8101d4c1904b7/pkg/ui/app/redux/cachedDataReducer.ts#L10)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I'm kind of surprised this worked.
</blockquote></details>

Typescript module syntax has a confusing history - this import has to both locate the typescript typings for moment *and* transpile to a statement that finds the "moment" module in the JS Module system being targeted.  The second part is especially tricky, since some module systems (such as CommonJS) allow modules to be constructed in a way that isn't fully compatible with ES6-style modules.

To get around this, typescript would allow you to do things like mix ES6-style imports with commonjs-style imports; the situations where you have to do this are weird, and in many cases (such as this one) were not actually necessary; or rather, perhaps it was necessary in the past, but our build system as evolved and now the ES6 style import works correctly for moment.

Basically, there are many layers at work here and occasionally you'll run into an odd situation where something doesn't immediately work as you expect.

---

*[pkg/ui/app/redux/nodes.ts, line 63 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kg-jehmFhBshWFjLD_c:-KgG3s2AYI41YrMG2KWE:b2u029x) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/redux/nodes.ts#L63)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary><blockquote>

I think a good pattern going forward would be to call the various accessor/selector functions something like `livenessesSelector` and then the argument `livenesses`.
</blockquote></details>

Yes, the linter does do this, it's considered a shadow.  Perhaps using "Selector" might be the best option, as max suggested.  I've changed every selector in this file to have a `Selector` suffix, and the parameters in various functions have been changed to use the same name without the suffix. 

---

*[pkg/ui/app/redux/nodes.ts, line 81 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kg08Gtv21l6obMSG0vV:-KgG44PlpMT4SGa3XTAU:b5dj1l0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/redux/nodes.ts#L81)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary><blockquote>

Man this probably would've saved me a lot of pain. Especially in my angular days.
</blockquote></details>

It's been on my radar for a while, if a solution is already available we should use it as soon as is practical; but not in this PR, of course. :)

---

*[pkg/ui/app/redux/nodes.ts, line 102 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-KfxJCTQ3GWGcbpDPdHd:-KgG4FEcuWK9l8qa2pmf:b-sn53n4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/redux/nodes.ts#L102)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary><blockquote>

Can't speak for @mrtracy but I can say my trust in js is so low I personally just opt to use `lodash` because I expect it'll be backwards compatible and fast. If the native implementation is fastest it'll just use that under the hood. Also it just makes operations like this read more consistently because `lodash` `each` and `map` operate similarly over arrays and objects.

I hadn't looked this up before but just to further support this argument: `lodash` is faster than native `map` but slower than a `for` loop....
https://jsperf.com/native-map-vs-lodash-map/21

</blockquote></details>

i typically just use Lodash in all cases when interacting with collections; in cases where the native solution is available and better, lodash uses the native method directly AFAIK.

---

*[pkg/ui/app/redux/nodes.ts, line 103 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kg1WOqhpPQz60XTvz6j:-KgG8wP7JKgsnwlZJW5P:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/redux/nodes.ts#L103)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary><blockquote>

This function could be a one liner
```
(ns) => ns.desc.node_id.toString()
```
</blockquote></details>

Done.

---

*[pkg/ui/app/redux/nodes.ts, line 199 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-Kg1XRqZrxOyQjWCOBUF:-KgG4Yi4Wq7FheV_fRx_:bpmdqlw) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/redux/nodes.ts#L199)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary><blockquote>

There's probably a more succinct way of doing this using weird new JS stuff but I can't think of anything off the top of my head.
</blockquote></details>

AFAIK this is already the succinct way, there are no member names in this object.

---

*[pkg/ui/app/redux/nodes.ts, line 211 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-KfxJxxrh4g1xDRMC03d:-KgGC0KhUksy2-FIp9mU:bkupods) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/redux/nodes.ts#L211)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary><blockquote>

Cool idea
</blockquote></details>

Extracted this to util/types.ts, along with a more general description of its usage.

---

*[pkg/ui/app/redux/nodes.ts, line 212 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258#-KfiEo-Bed9nMdBAAIXO:-KgG4e_wdBP9KZK7wfkp:bz8xmuz) ([raw file](https://github.com/cockroachdb/cockroach/blob/c11b926d801af06395d33e7f74eef48e2b421077/pkg/ui/app/redux/nodes.ts#L212)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can this be const, too? can it be inlined below?
</blockquote></details>

It cannot be inlined below, the typeof statement cannot be applied to a function call.  Will change to a `const`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14258)*
<!-- Sent from Reviewable.io -->
